nupic/research/frameworks/vernon/trainables.py

```diff
@@ -182,33 +274,48 @@ def _train(self):
             self._kill_workers()
             raise
 
-    def _should_stop(self):
+
+class RayActorHelperMethods:
+    def get_node_ip(self):
+        """Returns the IP address of the current node."""
+        return socket.gethostbyname(socket.gethostname())
+
+    def get_free_port(self):
+        """Returns free TCP port in the current node"""
+        return get_free_port()
+
+
+class DistributedTrainable(RemoteTrainableBase):
+    """
+    Use a multiple synchronized remote processes. With this trainable, there are
+    num_gpus + 1 processes total: the trainable process and the experiment
+    processes.
+
+    The experiment_class must implement the Experiment and
+    DistributedAggregation interfaces.
+    """
+
+    @classmethod
+    def default_resource_request(cls, config):
         """
-        Whether or not we should stop the experiment
+        Configure the cluster resources used by this experiment
         """
-        # Check if we should stop the experiment
-        stop_status = self.procs[0].should_stop.remote()
-        if ray_utils.check_for_failure([stop_status]):
-            return ray.get(stop_status)
-        else:
-            # Stop on failures
-            return True
+        num_gpus = config.get("num_gpus", 0)
+        num_cpus = config.get("num_cpus", 1)
 
-    def _save(self, _=None):
-        self.logger.debug(f"_save: {self._trial_info.trial_name}({self.iteration})")
-        # All models are synchronized. Just save the state of first model
-        with warn_if_slow("ImagenetExperiment.get_state.remote"):
-            return ray.get(self.procs[0].get_state.remote())
+        if num_gpus > 0:
+            # Assign extra CPUs for dataloaders
+            workers = config.get("workers", 0)
+            num_cpus = workers * num_gpus
 
-    def _restore(self, state):
-        # Restore the state to every process
-        state_id = ray.put(state)
-        ray.get([w.set_state.remote(state_id) for w in self.procs])
+        return Resources(cpu=0, gpu=0, extra_cpu=num_cpus, extra_gpu=num_gpus)
 
-        # Update current iteration using experiment epoch
-        self._iteration = ray.get(self.procs[0].get_current_epoch.remote())
+    def _extend_experiment_class(self, experiment_class):
+        class ExtendedExpClass(RayActorHelperMethods, experiment_class):
```

Comment:
The abbreviation Exp sounded a bit weird given in all other classes in vernon we don't refrain from using full names even if super long.